### ISSUE TITLE
Add more number types

### ISSUE THREAD

**Issue by oxinabox** [google-research/dex-lang#192]
Following up on #191  
It would be useful to have a few more number types.
In particular it would be useful to have Rational numbers and 128 bit integers.
Calculating grid-coefficients for higher-order finite differencing[using `Rational{Int128}` for intermidiate values before converting to `Float64` at the end](https://github.com/JuliaDiff/FiniteDifferences.jl/pull/65) gained me several orders of magnitude in accuracy;
and presumably has other applications.

And I know people working in astronomy like there rediculously high precision floats because of working at different scales.

I suggest:

- [ ] Rational numbers
- [ ] Int128
- [ ] BigInt ([GMP](https://gmplib.org/) equiv)
- [ ] BigFloat ([MPFR](https://www.mpfr.org/) equiv)

**Comment by dougalm** [comment]
Good idea. There's a question of whether to do these in the compiler or at the library level? Where there's no special hardware support I tend to lean towards doing things at the library level. What does Julia do?

**Comment by oxinabox** [comment]
All* types in Julia are at the library level.
Including [Float64](https://github.com/JuliaLang/julia/blob/5e8279050bfeac11ff7b6596e927d76e321dcc3c/base/boot.jl#L197-L214) etc.
Julia provides the `primative` keyword to create words-of-bits, 
which one then might use `llvmcall` (or `ccall`) to define the fundermental operations (that will in many cases have hardware support).
Some exceptions like `Int64` are not literally defined at the library level because of bootstrapping reasons.

The types I mentioned above all in the julia standard library.
In packages there are a bunch more like [ArbNumerics.jl](https://github.com/JeffreySarnoff/ArbNumerics.jl) and [BitIntegers.jl](https://github.com/rfourquet/BitIntegers.jl) and [BFloat16.jl](https://github.com/JuliaMath/BFloat16s.jl)


<sub>*not technically true but close enough. One can get into semantic arguments easily enough.</sub>
